src/lib.rs

```diff
@@ -466,3 +483,443 @@ where
+where
+    DB: BonsaiDatabase,
+    ChangeID: id::Id,
+{
+    /// Create a new revertible storage instance
+    pub fn new(
+        db: DB,
+        config: BonsaiStorageConfig,
+    ) -> Result<Self, BonsaiStorageError<DB::DatabaseError>> {
+        let kvdb = KeyValueDB::new(db, config.into(), None);
+        Ok(Self {
+            db: kvdb,
+            cache_storage_modified: HashMap::new(),
+        })
+    }
+
+    pub fn new_from_transactional_state(
+        db: DB,
+        config: BonsaiStorageConfig,
+        created_at: ChangeID,
+    ) -> Result<Self, BonsaiStorageError<DB::DatabaseError>> {
+        let kvdb = KeyValueDB::new(db, config.into(), Some(created_at));
+        Ok(Self {
+            db: kvdb,
+            cache_storage_modified: HashMap::new(),
+        })
+    }
+
+    /// Insert a new key/value in the storage, overwriting the previous value if it exists.
+    /// If the value already exists it will overwrite it.
+    ///
+    /// > Note: changes will not be applied until the next `commit`
+    pub fn insert(&mut self, identifier: &[u8], key: &BitSlice<u8, Msb0>, value: &[u8]) {
+        let key = bitslice_to_bytes(key);
+
+        self.cache_storage_modified.insert(
+            TrieKey::new(identifier, TrieKeyType::Flat, &key),
+            InsertOrRemove::Insert(value.to_vec()),
+        );
+    }
+
+    /// Remove a key/value in the storage
+    /// If the value doesn't exist it will do nothing
+    ///
+    /// > Note: changes will not be applied until the next `commit`
+    pub fn remove(&mut self, identifier: &[u8], key: &BitSlice<u8, Msb0>) {
+        let key = bitslice_to_bytes(key);
+
+        self.cache_storage_modified.insert(
+            TrieKey::new(identifier, TrieKeyType::Flat, &key),
+            InsertOrRemove::Remove,
+        );
+    }
+
+    /// Get a value in the storage.
+    pub fn get(
+        &self,
+        identifier: &[u8],
+        key: &BitSlice<u8, Msb0>,
+    ) -> Result<Option<Vec<u8>>, BonsaiStorageError<DB::DatabaseError>> {
+        let key = bitslice_to_bytes(key);
+        let key = TrieKey::new(identifier, TrieKeyType::Flat, &key);
+
+        match self.db.get(&key)? {
+            Some(value) => Ok(Some(value)),
+            None => Ok(None),
+        }
+    }
+
+    /// Checks if the key exists in the storage.
+    pub fn contains(
+        &self,
+        identifier: &[u8],
+        key: &BitSlice<u8, Msb0>,
+    ) -> Result<bool, BonsaiStorageError<DB::DatabaseError>> {
+        let key = bitslice_to_bytes(key);
+        self.db
+            .contains(&TrieKey::new(identifier, TrieKeyType::Flat, &key))
+    }
+
+    /// Go to a specific commit ID.
+    /// If insert/remove is called between the last `commit()` and a call to this function,
+    /// the in-memory changes will be discarded.
+    pub fn revert_to(
+        &mut self,
+        requested_id: ChangeID,
+    ) -> Result<(), BonsaiStorageError<DB::DatabaseError>> {
+        let kv = &mut self.db;
```

Comment:
Is it possible to make a function with this code and the one from revert_to (that is not tied to any of the storage) and call it here instead of duplicating the code ?

Reply:
Yep! I'm generally a bit on the fence when it comes to refactoring small bits of code like this, as I tend to prefer having functions explicitly state their inner workings, but I think in this case it's not really an issue as long as it stays in the same file :)